File: packages/sdk/src/conversation/middlewares/cardActionMiddleware.ts
import {
  ActivityTypes,
  CardFactory,
  InvokeResponse,
  MessageFactory,
  Middleware,
  StatusCodes,
  TurnContext,
} from "botbuilder";
import { AdaptiveCardResponse, TeamsFxAdaptiveCardActionHandler } from "../interface";
import { InvokeResponseFactory } from "../invokeResponseFactory";

/**
 * @internal
 */
export class CardActionMiddleware implements Middleware {
  public readonly actionHandlers: TeamsFxAdaptiveCardActionHandler[] = [];
  private readonly defaultMessage: string = "Your response was sent to the app";

  constructor(handlers?: TeamsFxAdaptiveCardActionHandler[]) {
    if (handlers && handlers.length > 0) {
      this.actionHandlers.push(...handlers);
    }
  }

  async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
    if (context.activity.name === "adaptiveCard/action") {
      const action = context.activity.value.action;
      const actionVerb = action.verb;

      for (const handler of this.actionHandlers) {
        if (handler.triggerVerb === actionVerb) {
          let response: InvokeResponse;
          try {
            response = await handler.handleActionInvoked(context, action.data);
          } catch (error: any) {
            const errorResponse = InvokeResponseFactory.errorResponse(
              StatusCodes.INTERNAL_SERVER_ERROR,
              error.message
            );
            await this.sendInvokeResponse(context, errorResponse);
            throw error;
          }

          const responseType = response.body?.type;
          switch (responseType) {
            case "application/vnd.microsoft.activity.message":
              await this.sendInvokeResponse(context, response);
              break;
            case "application/vnd.microsoft.card.adaptive":
              const card = response.body?.value;
              if (!card) {
                await this.sendInvokeResponse(
                  context,
                  InvokeResponseFactory.textMessage(this.defaultMessage)
                );
                throw new Error(`Adaptive card content cannot be found in the response body`);
              }

              if (card.refresh && handler.adaptiveCardResponse !== AdaptiveCardResponse.NewForAll) {
                // Card won't be refreshed with AdaptiveCardResponse.ReplaceForInteractor.
                // So set to AdaptiveCardResponse.ReplaceForAll here.
                handler.adaptiveCardResponse = AdaptiveCardResponse.ReplaceForAll;
              }

              const activity = MessageFactory.attachment(CardFactory.adaptiveCard(card));
              if (handler.adaptiveCardResponse === AdaptiveCardResponse.NewForAll) {
                await this.sendInvokeResponse(
                  context,
                  InvokeResponseFactory.textMessage(this.defaultMessage)
                );
                await context.sendActivity(activity);
              } else if (handler.adaptiveCardResponse === AdaptiveCardResponse.ReplaceForAll) {
                activity.id = context.activity.replyToId;
                await context.updateActivity(activity);
                await this.sendInvokeResponse(context, response);
              } else {
                await this.sendInvokeResponse(context, response);
              }
              break;
            case "application/vnd.microsoft.error":
            default:
              await this.sendInvokeResponse(context, response);
              break;
          }
        }
      }
    }

    await next();
  }

  private async sendInvokeResponse(context: TurnContext, response: InvokeResponse) {
    await context.sendActivity({
      type: ActivityTypes.InvokeResponse,
      value: response,
    });
  }
}

File: packages/dotnet-sdk/src/TeamsFx/Conversation/CardActionMiddleware.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace Microsoft.TeamsFx.Conversation
{
    using Microsoft.Bot.Builder;
    using Microsoft.Bot.Schema;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Linq;
    using System.Threading;
    using System.Threading.Tasks;

    internal class CardActionMiddleware : IMiddleware
    {
        internal readonly string defaultMessage = "Your response was sent to the app";

        /// <summary>
        /// Gets or sets the list of <see cref="IAdaptiveCardActionHandler"/> instances registered with the middleware.
        /// </summary>
        /// <value>
        /// The list of <see cref="IAdaptiveCardActionHandler"/> instances registered with the middleware.
        /// </value>
        public IList<IAdaptiveCardActionHandler> CardActionHandlers { get; private set; } = new List<IAdaptiveCardActionHandler>();

        /// <summary>
        /// Initializes a new instance of the <see cref="CommandResponseMiddleware"/> class.
        /// </summary>
        /// <param name="cardActionHandlers">A list of command handlers.</param>
        public CardActionMiddleware(IList<IAdaptiveCardActionHandler> cardActionHandlers = null)
        {
            if (cardActionHandlers != null && cardActionHandlers.Any())
            {
                CardActionHandlers = cardActionHandlers;
            }
        }

        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default)
        {
            if (turnContext.Activity.Name == "adaptiveCard/action")
            {
                var actionActivity = JsonConvert.DeserializeObject<AdaptiveCardInvokeValue>(turnContext.Activity.Value.ToString());
                string verb = actionActivity.Action.Verb;

                foreach (var handler in CardActionHandlers)
                {
                    if (handler.TriggerVerb.ToLowerInvariant() == verb?.ToString().ToLowerInvariant())
                    {
                        var invokeResponse = await handler.HandleActionInvokedAsync(turnContext, actionActivity.Action.Data, cancellationToken).ConfigureAwait(false);

                        if (invokeResponse != null)
                        {
                            var body = JObject.FromObject(invokeResponse.Body);
                            var responseValue = JsonConvert.DeserializeObject<AdaptiveCardInvokeResponse>(body.ToString());
                            string contentType = responseValue.Type;

                            if (contentType == InvokeResponseContentType.AdaptiveCard)
                            {
                                var card = responseValue.Value;
                                if (card == null)
                                {
                                    string errorMessage = "Adaptive card content cannot be null.";
                                    await SendInvokeResponseAsync(
                                        turnContext,
                                        InvokeResponseFactory.ErrorResponse(InvokeResponseErrorCode.InternalServerError,
                                        errorMessage), cancellationToken).ConfigureAwait(false);
                                    throw new ExceptionWithCode(errorMessage, ExceptionCode.InvalidParameter);
                                }

                                var isRefresh = ((JObject)card).ContainsKey("refresh");
                                var adaptiveCardResponse = handler.AdaptiveCardResponse;

                                if (isRefresh && handler.AdaptiveCardResponse == AdaptiveCardResponse.ReplaceForInteractor)
                                {
                                    // Ensure the base card for refresh action is able be viewed for all members
                                    // Otherwise, the refresh won't be triggered when user see the base card in Teams.
                                    adaptiveCardResponse = AdaptiveCardResponse.ReplaceForAll;
                                }

                                var messageActivity = MessageFactory.Attachment
                                (
                                    new Attachment
                                    {
                                        ContentType = "application/vnd.microsoft.card.adaptive",
                                        Content = card,
                                    }
                                );

                                switch (adaptiveCardResponse)
                                {
                                    case AdaptiveCardResponse.NewForAll:
                                        await SendInvokeResponseAsync(turnContext, InvokeResponseFactory.TextMessage(defaultMessage), cancellationToken).ConfigureAwait(false);
                                        await turnContext.SendActivityAsync(messageActivity, cancellationToken).ConfigureAwait(false);
                                        break;
                                    case AdaptiveCardResponse.ReplaceForAll:
                                        await SendInvokeResponseAsync(turnContext, invokeResponse, cancellationToken).ConfigureAwait(false);
                                        messageActivity.Id = turnContext.Activity.ReplyToId;
                                        await turnContext.UpdateActivityAsync(messageActivity, cancellationToken).ConfigureAwait(false);
                                        break;
                                    case AdaptiveCardResponse.ReplaceForInteractor:
                                        await SendInvokeResponseAsync(turnContext, invokeResponse, cancellationToken).ConfigureAwait(false);
                                        break;
                                }
                            }
                            else
                            {
                                // text message or error response
                                await SendInvokeResponseAsync(turnContext, invokeResponse, cancellationToken).ConfigureAwait(false);
                            }
                        }
 
                        break;
                    }
                }
            }

            await next(cancellationToken).ConfigureAwait(false);
        }

        private async static Task SendInvokeResponseAsync(ITurnContext context, InvokeResponse response, CancellationToken cancellationToken)
        {
            var invokeActivity = new Activity
            {
                Type = ActivityTypesEx.InvokeResponse,
                Value = response
            };

            await context.SendActivityAsync(invokeActivity, cancellationToken).ConfigureAwait(false);
        }
    }
}
